haystack/preview/components/fetchers/link_content.py

```diff
@@ -0,0 +1,225 @@
+        :param retry_attempts: The number of times to retry fetching content. Defaults to 2.
+        """
+        super().__init__()
+        self.raise_on_failure = raise_on_failure
+        self.user_agents = user_agents or [LinkContentFetcher._USER_AGENT]
+        self.current_user_agent_idx: int = 0
+        self.retry_attempts = retry_attempts or 2
+        self.handlers: Dict[str, Callable] = defaultdict(lambda: text_content_handler)
+
+        # register default content handlers that extract data from the response
+        self._register_content_handler("text/html", text_content_handler)
+        self._register_content_handler("text/plain", text_content_handler)
+        self._register_content_handler("application/pdf", binary_content_handler)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self,
+            raise_on_failure=self.raise_on_failure,
+            user_agents=self.user_agents,
+            retry_attempts=self.retry_attempts,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "LinkContentFetcher":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    def fetch(self, url: str, timeout: Optional[int] = 3, doc_kwargs: Optional[dict] = None) -> Document:
+        """
+        Fetches content from a URL and converts it to a Document objects. If no content is extracted,
+        an empty Document object is returned (if raise_on_failure is False).
+
+        :param url: URL to fetch content from.
+        :param timeout: Timeout in seconds for the request.
+        :param doc_kwargs: Optional kwargs to pass to the Document constructor.
+        :return: List of Document objects or an empty list if no content is extracted.
+        """
+        if not self._is_valid_url(url):
+            raise InvalidURL("Invalid or missing URL: {}".format(url))
+
+        doc_kwargs = doc_kwargs or {}
+        extracted_doc: Dict[str, Any] = {"metadata": {"url": url, "timestamp": int(datetime.utcnow().timestamp())}}
+        extracted_doc.update(doc_kwargs)
+        response = self._get_response(url, timeout=timeout or 3)
+        has_content = response.status_code == HTTPStatus.OK and (response.text or response.content)
+        if has_content:
+            content_type = self._get_content_type(response)
+            handler: Callable = self.handlers[content_type]
+            extracted_content = handler(response)
+            extracted_doc["content"] = extracted_content
+            # TODO assign content_type to created document
+            return Document(**extracted_doc)
+        else:
+            if self.raise_on_failure:
+                raise Exception(f"Couldn't retrieve content from {url}")
+            return Document(content="")
```

Comment:
It seems to me like we're managing exceptions in a very complicated way here.

`self._get_response` seems designed to handle all exceptions already, so why are we checking again for status codes and raising exceptions? Could we simplify this block into:

```suggestion
        response = self._get_response(url, timeout=timeout or 3)
        if response:
            content_type = self._get_content_type(response)
            handler: Callable = self.handlers[content_type]
            extracted_doc["content"] = handler(response)
            documents.append(Document(**extracted_doc))  # TODO assign content_type to created document
```

Note that I would also not return empty Documents in case of failure. Let's just skip over those if `raise_on_failure` is `False`.